### ISSUE TITLE
Try making PDFs more amenable to Apple Preview PDF Linearization

### ISSUE THREAD

**Issue by fduxiao** [typst/typst#4987]
### Description

I made a pdf with some math symbols. It displays well with a pdf reader, but when I printed it, all math symbols are missing.
It fails even for this simple example:
```typst
// test.typ
some math $x^2$
```

It works on a PDF reader.
<img width="335" alt="image" src="https://github.com/user-attachments/assets/a52c6651-765c-41c7-afdb-1319873b0cd4">

But fails if I print it.
![image](https://github.com/user-attachments/assets/ff38f2bd-345e-4683-a184-aeefe54d61a4)

I also tried pdf linearization. The result is even a blank pdf.
<img width="363" alt="image" src="https://github.com/user-attachments/assets/fb8e26b8-736b-42f3-8597-19b256cfed4d">

I finally solved this issue by turning it to postscript and back. But that will break all links in the document.
Is this related to the output format of pdf?

I am using Typst 0.11.1 installed by Homebrew.

### Operating system

macOS

### Typst version

- [X] I am using the latest version of Typst

**Comment by LaurenzV** [comment]
There have been printer issues in the past, but this PR should have fixed them: https://github.com/typst/typst/pull/4373

So if possible, feel free to either try building Typst yourself from `main`, or wait for the 0.12 release.

**Comment by fduxiao** [comment]
@LaurenzV Thanks for the help. I tried compiling from source. It works well for the printer, but linearization is still empty, except some drawing (`#line`).

**Comment by LaurenzV** [comment]
What do you mean by linearization? Creating a linearized PDF file? Sounds like the problem is then with whatever software you are using to create the linearized file...

**Comment by fduxiao** [comment]
@LaurenzV Yes, I am using the default pdf viewer (Preview) on macOS to linearize the pdf. The problem is that if you send the paper to a publisher, then an editor may reject it and tell you to use a pdf that is compatible with their pdf software so they can linearize it. This is not the case for latex.

**Comment by LaurenzV** [comment]
Indeed, but if Apple Preview is the software that is breaking the PDF, then there isn't really much we can do about this... Apart from supporting the creation of linearized PDFs out-of-the-box. So if that's what you want, that's a valid feature request, but then I would rename this issue to reflect that.

I just tried this myself, and I can indeed reproduce it, after linearizing the file with Apple Preview it doesn't work anymore, but it's because Apple Preview creates an invalid file, so it's a bug on their side... If I linearize it using mupdf by invoking `mutool clean -l file.pdf`, the output works fine, so if it's really important for you to have a linearized PDF, I suggest using that tool.

**Comment by fduxiao** [comment]
It won't be a problem for me. My concern is that there could be some potential incompatibility with existing tools. Different publishers may have different tools for pdf manipulation. If they find that a Typst compiled pdf is not working on their tools, they won't change their `Apple Preview` to `mutual`, but ask you to use latex instead of Typst. I am not requesting a feature for linearized pdf, but reporting that a Typst compiled pdf behaves differently from a latex compiled one.

**Comment by Enivex** [comment]
> It won't be a problem for me. My concern is that there could be some potential incompatibility with existing tools. Different publishers may have different tools for pdf manipulation. If they find that a Typst compiled pdf is not working on their tools, they won't change their `Apple Preview` to `mutool`, but ask you to use latex instead of Typst. I am not requesting a feature for linearized pdf, but reporting that a Typst compiled pdf behaves differently from a latex compiled one.

It wouldn't be the first issue with Apple Preview. As long as typst follows the pdf standard, the blame is on the reader.

Anyway, this issue seems to be resolved on main, so I'll close it.

**Comment by laurmaedje** [comment]
Do we know what Preview is messing up? In theory, it could also mean a bug in Typst that other viewers are able to recover from. Even if improbable it would be good to know for sure.

**Comment by Enivex** [comment]
> Do we know what Preview is messing up? In theory, it could also mean a bug in Typst that other viewers are able to recover from. Even if improbable it would be good to know for sure.

It would be hard to figure out when Preview is a black box. I guess an educated guess could be made by looking at exactly what it spits out.

**Comment by LaurenzV** [comment]
> Do we know what Preview is messing up? In theory, it could also mean a bug in Typst that other viewers are able to recover from. Even if improbable it would be good to know for sure.

Before:
```
9 0 obj
<<
  /Type /Font
  /Subtype /Type0
  /BaseFont /ZBDUEJ+NewCMMath-Book-Identity-H
  /Encoding /Identity-H
  /DescendantFonts [ 10 0 R ]
  /ToUnicode 12 0 R
>>
endobj

10 0 obj
<<
  /Type /Font
  /Subtype /CIDFontType0
  /BaseFont /ZBDUEJ+NewCMMath-Book
  /CIDSystemInfo <<
    /Registry (Adobe)
    /Ordering (Identity)
    /Supplement 0
  >>
  /FontDescriptor 11 0 R
  /DW 0
  /W [ 0 0 500 2712 2712 572 4598 4598 569 ]
>>
endobj

11 0 obj
<<
  /Type /FontDescriptor
  /FontName /ZBDUEJ+NewCMMath-Book
  /Flags 131076
  /FontBBox [ -1044 -3060 4087.9998 3560 ]
  /ItalicAngle 0
  /Ascent 806
  /Descent -194
  /CapHeight 683
  /StemV 95.4
  /FontFile3 13 0 R
>>
endobj
```

After:
```
11 0 obj
<<
  /Type /Font
  /Subtype /Type0
  /Encoding /Identity-H
  /DescendantFonts [ 16 0 R ]
  /BaseFont /AAAAAC+NewCMMath-Book
>>
endobj

16 0 obj
<<
  /Type /FontDescriptor
  /FontName /AAAAAB+LinLibertine
  /Flags 4
  /FontBBox [ -1082 -256 6171 1125 ]
  /ItalicAngle 0
  /Ascent 894
  /Descent -246
  /CapHeight 658
  /StemV 0
  /XHeight 429
  /AvgWidth 566
  /MaxWidth 6272
  /FontFile2 17 0 R
>>
endobj
```

**Comment by LaurenzV** [comment]
It completely messes up the fonts...

```
10 0 obj
<<
  /Type /Font
  /Subtype /Type0
  /Encoding /Identity-H
  /DescendantFonts [ 11 0 R ]
  /BaseFont /AAAAAB+LinLibertine
  /ToUnicode 14 0 R
>>
endobj

11 0 obj
<<
  /Type /Font
  /Subtype /Type0
  /Encoding /Identity-H
  /DescendantFonts [ 16 0 R ]
  /BaseFont /AAAAAC+NewCMMath-Book
>>
endobj
```

**Comment by LaurenzV** [comment]
I just created the most simple Latex document in Overleaf:

```
\documentclass{article}
\begin{document}
Hi.
\end{document}
```

and tried the same thing, and the resulting file is also broken... So Latex has the same problem, and it's just Apple Preview's feature being incredibly broken. :)

**Comment by LaurenzV** [comment]
Okay, update: In pdftex, it does work, but not in LuaLaTex. I suspect that whatever issue Apple Preview is having, only happens when using CID-keyed fonts... But I'm not sure we can really change this.

**Comment by LaurenzV** [comment]
On the other hand, Word-generated documents seem to work fine... I guess let's reopen the issue, maybe there is something we can do to circumvent it, but let's adapt the title accordingly.

**Comment by ghost** [comment]
I guess it could be related to the math font format (CFF vs. TTF) like shown [here](https://github.com/typst/typst/issues/5392#issuecomment-2466259221).

Regarding "linearization": I don't think that this is related to Typst. It is rather an Apple Preview bug since implementing Linearization correctly is a difficult task and we know that Apple Preview has hundreds of hard bugs.
To make sure that the linearized PDF is actually valid, I would first check it with a PDF validator.

**Comment by LaurenzV** [comment]
Yeah, I think we can just close this, not worth keeping this open since it's seemingly an Apple Preview bug.
